File: server/webapp/WEB-INF/rails.new/app/presenters/api_v1/version_info_representer.rb
##########################################################################
# Copyright 2015 ThoughtWorks, Inc.
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
##########################################################################

module ApiV1
  class VersionInfoRepresenter < ApiV1::BaseRepresenter
    def initialize(version_info, system_environment)
      @version_info = version_info
      @system_environment = system_environment
      super(@version_info)
    end

    link :doc do
      'https://api.go.cd/#versionInfo'
    end

    link :self do |opts|
      opts[:url_builder].apiv1_stale_version_info_url
    end

    property :getComponentName, as: :component_name
    property :update_server_url, exec_context: :decorator
    property :installed_version, exec_context: :decorator
    property :latest_version, exec_context: :decorator

    def installed_version
      @version_info.getInstalledVersion.toString
    end

    def latest_version
      @version_info.getLatestVersion.blank? ? nil : @version_info.getLatestVersion.to_string
    end

    def update_server_url
      uri = URI(@system_environment.getUpdateServerUrl)
      ar = URI.decode_www_form(uri.query || "") << ['current_version', @version_info.getInstalledVersion.toString || 'unknown']
      uri.query = URI.encode_www_form(ar)
      uri.to_s
    end
  end
end

File: api/api-version-infos-v1/src/main/java/com/thoughtworks/go/apiv1/versioninfos/representers/VersionInfoRepresenter.java
/*
 * Copyright 2021 ThoughtWorks, Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.thoughtworks.go.apiv1.versioninfos.representers;

import com.thoughtworks.go.api.base.OutputWriter;
import com.thoughtworks.go.domain.VersionInfo;
import com.thoughtworks.go.spark.Routes;
import com.thoughtworks.go.util.SystemEnvironment;
import org.apache.http.NameValuePair;
import org.apache.http.client.utils.URIBuilder;
import org.apache.http.message.BasicNameValuePair;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import java.net.URI;
import java.net.URISyntaxException;
import java.util.List;

import static com.thoughtworks.go.CurrentGoCDVersion.apiDocsUrl;

public class VersionInfoRepresenter {
    private static final Logger LOG = LoggerFactory.getLogger(VersionInfoRepresenter.class);

    public static void toJSON(OutputWriter outputWriter, VersionInfo versionInfo, SystemEnvironment systemEnvironment) {
        if (versionInfo == null) {
            return;
        }
        outputWriter.addLinks(linkWriter -> {
            linkWriter.addAbsoluteLink("doc", apiDocsUrl("#version-info"))
                    .addLink("self", Routes.VersionInfos.BASE + Routes.VersionInfos.STALE);
        }).add("component_name", versionInfo.getComponentName());

        renderUpdatedServerUrl(outputWriter, versionInfo, systemEnvironment);

        outputWriter.add("installed_version", versionInfo.getInstalledVersion().toString());

        if (versionInfo.getLatestVersion() == null) {
            outputWriter.renderNull("latest_version");
        } else {
            outputWriter.add("latest_version", versionInfo.getLatestVersion().toString());
        }
    }

    private static void renderUpdatedServerUrl(OutputWriter outputWriter, VersionInfo versionInfo, SystemEnvironment systemEnvironment) {
        try {
            URI updateServerUri = URI.create(systemEnvironment.getUpdateServerUrl());
            URIBuilder uriBuilder = new URIBuilder(updateServerUri);

            String currentVersion = versionInfo.getInstalledVersion() != null ? versionInfo.getInstalledVersion().toString() : "unknown";
            uriBuilder.addParameter("current_version", currentVersion);

            String updatedServeUrl = uriBuilder.build().toString();
            outputWriter.add("update_server_url", updatedServeUrl);
        } catch (Exception e) {
            LOG.error("Exception occurred while building updated server url", e);
            outputWriter.renderNull("update_server_url");
        }
    }
}
